platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java

```diff
@@ -55,12 +67,20 @@ public Set<Attribute<?>> keySet() {
 
     @Override
     public <T> AttributeContainer attribute(Attribute<T> key, T value) {
-        throw new IllegalArgumentException(String.format("Cannot change attributes of dependency %s after it has been resolved", owner.getDisplayName()));
+        if (delegate instanceof ImmutableAttributes) {
+            throw new IllegalArgumentException(String.format("Cannot change attributes of dependency %s after it has been resolved", owner.getDisplayName()));
```

Comment:
🤔 I think `IllegalStateException` is more appropriate here.  Also, putting at least the attribute name (possibly value?) in the error message would probably be helpful to the debugger who encounters this and tries to understand the behavior.

Why does this say "attributes of dependency X"?  No guarantee it's a dependency.  I think it would be better to be more generic "attributes associated with X"

Reply:
Not sure why the text was like this before. I've reworded the exception and updated it to `IllegalStateException`